### PR TITLE
[MM-39548] Allow end-users to fetch marketplace plugins, but exclude information on server-only plugins

### PR DESCRIPTION
**Pull request by mickmister**
#### Summary

The purpose of this PR, is so that the App Bar can access the plugin icon data for webapp plugins that register a channel header component. This makes it so the logos for different plugins are styled correctly with the right branding, following the designs of the App Bar. On the client-side, we will look for the marketplace icon, and fall back to the provided channel header React component if no icon data is available.

This PR makes it so non-sysadmins can fetch marketplace plugin information, but it excludes server-only plugins. It was determined in a [discussion](https://community-daily.mattermost.com/core/pl/3b14ptmzf7yotxjg3nyt15y47e) that the read permissions for installed plugin information are in place for the case of an admin wanting to keep the presence of a plugin be hidden. The linked post notes that server-only plugins are the main issue for this situation.

I thought of just emptying the `InstalledVersion` for each plugin in the response, but custom plugins will show up in the response that the admin may have wanted to keep hidden. We don't have any flag for "this plugin came from the marketplace" or is a custom plugin.

#### Ticket Link

Fixes https://mattermost.atlassian.net/browse/MM-39548

#### Related Pull Requests

https://github.com/mattermost/mattermost-webapp/pull/9273

#### Release Note

```release-note
NONE
```

### CODE REVIEW

**Review by agarciamontoro — APPROVED** [review]
Code looks good!

File: api4/plugin.go
Comment (agarciamontoro) [reply api4/plugin.go:286]: Can't this info be stored in the redux store, calling the endpoint just once?

Comment (mickmister) [reply api4/plugin.go:286]: @agarciamontoro This endpoint will be called once on mount of the App Bar. At the moment, this happens whenever we visit the Channels product. The component then pulls from the redux store to get the result stored there.

Comment (agarciamontoro) [reply api4/plugin.go:286]: Then what's the performance concern?

Comment (hanzei) [reply api4/plugin.go:286]: The concern is that every full redux freshed (e.g. login) will trigger an outgoing HTTP request

**Review by hanzei — COMMENTED** [review]
The current approach still allows clients to trigger an outgoing http request to the Marketplace. Is the remote Marketplace data needed for client? Or are the `local` clients enough?

**Comment by mickmister** [comment]
> The current approach still allows clients to trigger an outgoing http request to the Marketplace. Is the remote Marketplace data needed for client? Or are the `local` clients enough?

@hanzei I've changed the frontend PR to use the `local_only` parameter. Though we are still allowing non-admins to manually form a request that will contact the remote marketplace server. My thinking was that the remote marketplace is public and can be arbitrarily contacted via HTTP, so anyone can contact the marketplace server manually anyway. Outgoing requests from the server is probably the reason you brought this up though right? Not because of the remote marketplace itself.

How about I change this PR to make it so we _force_ `local_only` when the request is coming from a non-admin?

File: api4/plugin_test.go
Comment (mickmister) [inline api4/plugin_test.go:924]: This is written this way, because when the `local_only` param is used, an extra label is put on each plugin saying that the plugin is not available in the marketplace. So the structs arent directly equal past the `BaseMarketplacePlugin`

**Comment by mickmister** [comment]
I've changed the PR to enforce on the backend non-sysadmins to use `local_true`

**Review by hanzei — APPROVED** [review]
Awesome :tada:

**Comment by mickmister** [comment]
@DHaussermann Wondering your thoughts on merging this now to make it easier to test https://github.com/mattermost/mattermost-webapp/pull/9273. This change was introduced so non-sysadmins can access the marketplace icons for plugins that display in the App Bar, whereas previously the marketplace endpoint was restricted to sysadmins only.

**Comment by mickmister** [comment]
@hanzei Thinking about this some more, it may be wasteful to fetch all of the webapp plugin icons every time we want to calculate the icons for the App Bar items. Consider this case:

The user refreshes their page. Marketplace icons are fetched for each channel header component that is to be placed in the App Bar. Another webapp plugin is enabled. We need the marketplace icon for this plugin, so _all_ users now need to fetch the marketplace icon for this plugin. This results in all users fetching the icons for all installed plugins.

Keep in mind that these icons are delivered as data URLs, so the image content is calculated every time someone requests these for the `local_only` case:

https://github.com/mattermost/mattermost-server/blob/12dc171a606e20be2d10b7d7f2fad764a89b726c/app/plugin.go#L683-L689

I'm thinking we could instead do one of the following:

- Expose an endpoint that allows clients to fetch the image for a given webapp plugin
- Include a new field in the webapp manifest we deliver to the client, that contains this `icon_data` data URL

Adding a new endpoint to fetch the icon for a given webapp plugin is how I originally was going to implement this, but I thought adding a new endpoint for this specific use case would be overkill. Though now I'm thinking constantly fetching the images for all App Bar icons is what is overkill (performance-wise in this case), since there's no opportunity for client-side caching of these images. I suppose we would include the version of the plugin as a query parameter to enforce proper client-side caching.

I also like this proposed solution because there's no extra API call the client needs to keep track of, it's just fetching the images using the native `img` tag functionality. If the image fails to load for whatever reason, we can fall back to the atual channel header icon as we are currently doing in the webapp PR for this.

**Comment by hanzei** [comment]
@mickmister Could you please elaborate why the web clients need the Marketplace plugins in the first place? In https://docs.google.com/document/d/1-49BDN_1TN008_9F3IDR0TEHhIypcGxhG0akqzZer8o/edit# it is mentioned, that the channel header icon should be used for the app bar. Is there some kind of technical overview on the app bar feature?

**Comment by mickmister** [comment]
> @mickmister Could you please elaborate why the web clients need the Marketplace plugins in the first place? In https://docs.google.com/document/d/1-49BDN_1TN008_9F3IDR0TEHhIypcGxhG0akqzZer8o/edit# it is mentioned, that the channel header icon should be used for the app bar.

@hanzei It was discussed in our weekly App Bar meeting that we are instead going with using the marketplace icons. I've just added you to the participants of this meeting. That document was not updated to reflect this decision.

> Is there some kind of technical overview on the app bar feature?

I believe that document you linked is the closest thing to a technical overview at the moment, which is obviously not its purpose.

**Comment by mickmister** [comment]
@hanzei The goal for using the marketplace icons is so we can have the individual Apps use their own branding, instead of having plain gray icons in the App Bar. cc @matthewbirtch

**Comment by hanzei** [comment]
@mickmister Is there a need to fetch the _Marketplace_ plugins or is the list of local plugins enough? If it's the latter, we could extend the `api/v4/plugins` endpoint to also include the `icon_data`. To be clear: The icons aren't Marketplace ones. They come from the plugins them self. The Marketplace Modal only uses them.

**Comment by mickmister** [comment]
@hanzei Technically this will only fetch local plugins because the webapp is using `local_only=true` for its request for the app bar icons. So the marketplace is actually not used. Anyway, as I said offline I'm wanting to include the icon in the API endpoint that retrieves info on webapp plugins.

Copying my thoughts here to discuss publicly:

I've thought of another plan here. The `plugins/webapp` endpoint returns a `*Manifest` that has its values filtered for client-only use, using `manifest.ClientManifest() *Manifest`. 
[mattermost/mattermost-server/model/manifest.go](https://github.com/mattermost/mattermost-server/blob/009d02c57a923769dcb77acc32e57923f1f1d8f7/model/manifest.go#L233)


I'm thinking we can have a `ManifestWithIcon` struct that this endpoint returns instead. This new struct would also be delivered in the `plugin_enabled` websocket event. This way no new API calls are added, and the client simply has the icons from data it's already storing client-side.

```go
type ManifestWithIcon struct {
    *Manifest
    IconData string `json:"icon_data"`
}

func (m *Manifest) ClientManifestWithIcon() *ManifestWithIcon {
```

**Comment by mickmister** [comment]
@hanzei As you pointed out in the document you linked, the original decision was to go with the fallback channel header icons. Since then, the general consensus changed to use the marketplace icons, and the document was not updated. I apologize for all the miscommunication around this.

@hanzei @matthewbirtch In order to get the first phase of this feature completed sooner, I'm thinking we should go with the original plan of using the fallback icons. Some factors to consider:
- It's still undecided how the icons should be delivered to the client, i.e. marketplace endpoint or another mechanism
- We're still working on the images that need to be used for each plugin
- We will need to create releases in each of these plugins once the icons are approved
- This feature is hidden by a feature flag, so we have control over which servers are using the feature.

If we go with using the fallback icons, then the focus will shift to ensuring all the known fallback icons are presentable. Here's the current state, with the Microsoft Teams Meetings one being the worst out of the collection at the top:

**Comment by matthewbirtch** [comment]
> @hanzei @matthewbirtch In order to get the first phase of this feature completed sooner, I'm thinking we should go with the original plan of using the fallback icons. Some factors to consider:
> 
> * It's still undecided how the icons should be delivered to the client, i.e. marketplace endpoint or another mechanism
> * We're still working on the images that need to be used for each plugin
> * We will need to create releases in each of these plugins once the icons are approved
> * This feature is hidden by a feature flag, so we have control over which servers are using the feature.
> 
> If we go with using the fallback icons, then the focus will shift to ensuring all the known fallback icons are presentable. Here's the current state, with the Microsoft Teams Meetings one being the worst out of the collection at the top:

I don't fully understand all the issues here, to be honest. The App Bar PR you were working on, @mickmister, seemed to handle both scenarios pretty well. I don't really get why we can't proceed that way.

To make it easiest for app/plugin developers, we should make it so they only have to have one icon image file uploaded. Ideally, that image file could be utilized for the marketplace, app bar and everywhere else the icon would be displayed. Is this not an approach we can proceed with?

**Comment by mickmister** [comment]
@matthewbirtch I'm really just trying to unblock the PR so it can be merged sooner. I'd rather have the plugin icons 

> It's still undecided how the icons should be delivered to the client, i.e. marketplace endpoint or another mechanism

@hanzei and I are meeting later today to discuss this. The issue is that we may be overcomplicating the usage of API endpoints in this PR by using the marketplace endpoint to fetch these icons. No matter what solution we go with in this regard, the images will still ultimately be sourced from the marketplace as you and I have been planning around thus far. So nothing will change for the process of delivering the icons through new plugin releases.

> We're still working on the images that need to be used for each plugin

As noted in https://github.com/mattermost/mattermost-marketplace/pull/35#issuecomment-981561845, the icons must be in SVG format. At the moment, we've been going in the direction of using PNGs for the new icons, so I think we'll need to pivot and use SVGs instead. Both server code and marketplace code has this assumption hardcoded, so it will likely stay the case that they need to be in SVG format unless there is a strong reason that combats the original reason that introduced the SVG requirement.

> This feature is hidden by a feature flag, so we have control over which servers are using the feature.

I mention this because we can deliver a suboptimal feature (using the fallback icons) for our first release of the feature, since we can conditionally have it enabled only for community for the time being. This is mainly so we can get the PR merged and focus on moving forward with the App Bar project, including getting the icons properly integrated with the component. I'm thinking that we should be using the marketplace icons with up-to-date plugin releases when we remove the feature flag logic.

**Comment by matthewbirtch** [comment]
> @matthewbirtch I'm really just trying to unblock the PR so it can be merged sooner. I'd rather have the plugin icons
> 
> > It's still undecided how the icons should be delivered to the client, i.e. marketplace endpoint or another mechanism
> 
> @hanzei and I are meeting later today to discuss this. The issue is that we may be overcomplicating the usage of API endpoints in this PR by using the marketplace endpoint to fetch these icons. No matter what solution we go with in this regard, the images will still ultimately be sourced from the marketplace as you and I have been planning around thus far. So nothing will change for the process of delivering the icons through new plugin releases.
> 
> > We're still working on the images that need to be used for each plugin
> 
> As noted in [mattermost/mattermost-marketplace#35 (comment)](https://github.com/mattermost/mattermost-marketplace/pull/35#issuecomment-981561845), the icons must be in SVG format. At the moment, we've been going in the direction of using PNGs for the new icons, so I think we'll need to pivot and use SVGs instead. Both server code and marketplace code has this assumption hardcoded, so it will likely stay the case that they need to be in SVG format unless there is a strong reason that combats the original reason that introduced the SVG requirement.
> 
> > This feature is hidden by a feature flag, so we have control over which servers are using the feature.
> 
> I mention this because we can deliver a suboptimal feature (using the fallback icons) for our first release of the feature, since we can conditionally have it enabled only for community for the time being. This is mainly so we can get the PR merged and focus on moving forward with the App Bar project, including getting the icons properly integrated with the component. I'm thinking that we should be using the marketplace icons with up-to-date plugin releases when we remove the feature flag logic.

@mickmister thanks for the detailed explanations. Maybe we should queue this up for our sync meeting or schedule a separate meeting to discuss this. The one worry I have for limiting to just SVGs is that not everyone has the tools to author an SVG, while creating JPGs or PNGs is much easier for most users. I think it's an obstacle we can overcome, but we need to provide clear guidelines and templates that export as an SVG.

**Comment by mickmister** [comment]
@DHaussermann @hanzei Should we close this PR? We are going against using the marketplace icons in the first phase of the App Bar, and we may not use this exact solution later anyway

**Comment by hanzei** [comment]
:+1: for closing the PR for now

**Pull request closed without merging**
